File: week 3/dot_product/dot_product.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using std::vector;

long long min_dot_product(vector<int> a, vector<int> b) {
  // write your code here
  long long result = 0;
  sort(a.begin(), a.end());
  sort(b.begin(), b.end());
  for (size_t i = 0; i < a.size(); i++) {
    result += (long long)a[i] * b[a.size()-i-1];
  }
  return result;
}

int main() {
  size_t n;
  std::cin >> n;
  vector<int> a(n), b(n);
  for (size_t i = 0; i < n; i++) {
    std::cin >> a[i];
  }
  for (size_t i = 0; i < n; i++) {
    std::cin >> b[i];
  }
  std::cout << min_dot_product(a, b) << std::endl;
}

File: week 3/dot_product/DotProduct.java
import java.util.*;

public class DotProduct {	
    private static long minDotProduct(int[] a, int[] b) {
		//write your code here
        long result = 0;
		Arrays.sort(a);	
		Arrays.sort(b);
		
        for (int i = 0; i < a.length; i++) {
            result += (long)a[i] * b[a.length-i-1];
        }
        return result;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
            a[i] = scanner.nextInt();
        }
        int[] b = new int[n];
        for (int i = 0; i < n; i++) {
            b[i] = scanner.nextInt();
        }
        System.out.println(minDotProduct(a, b));
    }
}
